### PR TITLE
Migrate setup.py to pyproject.toml

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- Migrate setup.py to pyproject.toml

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,9 +1,77 @@
...
+requires-python = ">=3.9"
+dynamic = ["version"]
+classifiers = [
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+  "Development Status :: 4 - Beta",
+  "Intended Audience :: Science/Research",
+  "Intended Audience :: System Administrators",
```
Comment (DanielYang59) [inline pyproject.toml:23]: This looks a bit off, why is `matminer` intended for `system admins`?

Comment (JaGeo) [reply pyproject.toml:23]: Just change if you think this doesn't fit.

Comment (DanielYang59) [reply pyproject.toml:23]: Thanks for the comment, I would remove it for now unless it's intended for other purposes @ml-evs

Comment (ml-evs) [reply pyproject.toml]: No idea, I'm just a caretaker

Comment (ml-evs) [reply pyproject.toml]: Don't think trove classifiers are that load-bearing though

File: pyproject.toml [hunk 2]
```diff
@@ -1,9 +1,76 @@
...
+    "coverage",
+    "coveralls",
+    "flake8",
+    "black",
+    "pylint",
+    "sphinx"
+]
+tests = ["ujson", "jmespath", "httplib2", "ase", "jsonschema",
+  "dscribe~=2.1", "mdf-forge", "aflow", "citrination-client"
+]
```
Comment (ml-evs) [inline pyproject.toml:63]: Might be cleaner to do

```suggestion
tests = ["matminer[mpds,describe,mdfforge,aflow,citrine,dev]"]
```

Comment (DanielYang59) [reply pyproject.toml]: Beautiful! I wasn't aware of this even exists! This is what I was looking for

**Comment by DanielYang59** [comment]
No problem, thanks for reviewing and the new trick :)

**Pull request merged**